File: Python/maybe.py
from typing import TypeVar, Generic, Callable

T = TypeVar('T')  # Define type variable "T"
U = TypeVar('U')  # Define type variable "U"

class Maybe(Generic[T]):
    def __init__(self, value: T | None = None):
        self.value = value

    def empty() ->"Maybe[T]":
        return Maybe()
        
    def pure(val: T) -> "Maybe[T]":
        return Maybe(val)

    def fmap(self, f: Callable[[T], U]) -> "Maybe[U]":
        return self if self.value is None else Maybe.pure(f(self.value))

    def and_then(self, f: Callable[[T], U]) -> "Maybe[U]":
        return self if self.value is None else f(self.value)

    def or_else(self, els) -> "Maybe[T]":
        return self if not self.value is None else els() if callable(els) else els

    def __or__(self, els) -> "Maybe[T]":
        return self.or_else(els)
        
    # def __mul__(self, m: Callable[[], Maybe[U]]) -> "Maybe[U]":
    def __mul__(self, m) -> "Maybe[U]":
        return self.and_then(lambda f: (m() if callable(m) else m).fmap(f))

File: Swift/Sources/Parser/Maybe.swift
struct Maybe<A> {
	var value: Optional<A>
	
	init(){
		self.value = Optional<A>.none
	}
	
	init(_ value: A){
		self.value = value
	}
	
	static func Nothing() -> Maybe<A>{
		return Maybe<A>()
	}
	
	static func Just(_ value: A) -> Maybe<A>{
		return Maybe<A>(value)
	}

	func map<B>(_ f: (A) -> B) -> Maybe<B> {
		if let v: A = value {
			return Maybe<B>.Just(f(v))
		} else { return Maybe<B>.Nothing() }
	}

	func flatMap<B>(_ f: (A) -> Maybe<B>) -> Maybe<B>{
		if let v: A = value {
			return f(v)
		} else { return Maybe<B>.Nothing() }
	}

	func orElse(_ f: () -> Maybe<A>) -> Maybe<A>{
		if let _: A = value {
			return self
		} else { return f() }
	}
}
